Chap_API_Proc_Mgmt.tex

```diff
@@ -607,10 +607,69 @@ \section{Process Locality}
 
 The relative locality of processes is often used to optimize their interactions with the hardware and other processes. \ac{PMIx} provides a means by which the host environment can communicate the locality of a given process using the \refapi{PMIx_generate_locality_string} to generate an abstracted representation of that value. This provides a human-readable format and allows the client to parse the locality string with a method of its choice that may differ from the one used by the server that generated it.
 
+There are times, however, when relative locality and other \ac{PMIx}-provided
+information doesn't include some element required by the application. In these
+instances, the application may need access to the full description of the
+local hardware topology. \ac{PMIx} does not itself generate such descriptions
+- there are multiple third-party libraries that fulfill that role. Instead,
+\ac{PMIx} offers an abstraction method by which users can obtain a pointer to
+the description. This transparently enables support for different methods of
+sharing the topology between the host environment (which may well have already
+generated it prior to local start of application processes) and the clients -
+e.g., through passing of a shared memory region.
 
 %%%%%%%%%%%
-\subsection{\code{PMIx_get_relative_locality}}
-\declareapi{PMIx_get_relative_locality}
+\subsection{\code{PMIx_Load_topology}}
+\declareapi{PMIx_Load_topology}
+
+%%%%
+\summary
+
+Load the local hardware topology description
+
+%%%%
+\format
+
+\versionMarker{4.0}
+\cspecificstart
+\begin{codepar}
+pmix_status_t
+PMIx_Load_topology(pmix_topology_t *topo);
+\end{codepar}
+\cspecificend
+
+\begin{arglist}
+\arginout{topo}{Address of a \refstruct{pmix_topology_t} structure where the topology information is to be loaded (handle)}
+\end{arglist}
+
+Returns \refconst{PMIX_SUCCESS}, indicating that the \refarg{topo} was successfully loaded, or an appropriate \ac{PMIx} error constant.
+
+%%%%
+\descr
+
+Obtain a pointer to the topology description of the local node. If the
+\refarg{source} field of the provided \refstruct{pmix_topology_t} is set, then
+the \ac{PMIx} library must return a description from the specified
+implementation or else indicate that the implementation is not available by
+returning the \refconst{PMIX_ERR_NOT_SUPPORTED} error constant.
+
+The returned pointer may point to a shared memory region or an actual instance
+of the topology description. In either case, the description shall be treated
+as a "read-only" object - attempts to modify the object are likely to fail and
+return an error. The \ac{PMIx} library is responsible for reference counting
+the returned object and performing any required cleanup either when the
+reference count reaches zero or the library itself finalizes.
```

Comment:
How would a PMIx library go about reference counting the returned object? We could increment the reference count when giving it to the client, but we don't have a 'release' function for the client to indicate that they are done with it. Maybe we just need a `PMIX_TOPO_RELEASE` call - and a sentence here like "when the returned object is no longer required then the caller must pass it to `PMIX_TOPO_RELEASE` to ensure safe cleanup."

Reply:
In thinking about it, we don't need this at all. We can just release the topology during finalize since that is already protected by internal refcount. We can internally track the source of the topology (i.e., whether it came from shmem/xml or was independently discovered) and take the appropriate action.

Reply:
see what you think of the latest commit - if all looks good, I'll rebase it into one commit and commit it.

Reply:
That sounds good